### PR TITLE
feat: Introduce the Discover dataset

### PR DESCRIPTION
**Pull request by lynnagara**
This is an experimental dataset that simply routes a query to either
the events or transactions datasets. It maps common columns to the
same name/type and fills non common columns with null.

This is an intermediate step that allows us to move the dataset
detection logic that switches between the events and transactions table
from Sentry into Snuba. In future, this will likely be solved via a
materialized view instead.

### CODE REVIEW

File: snuba/datasets/all_events.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return get_dataset(EVENTS) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+
+
+class AllEventsDataset(TimeSeriesDataset):
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: I know naming things is hard, though, keep in mind, that changing dataset names tends to be annoyingly hard.
Let's be sure this is the name we want to use long term.

File: snuba/datasets/all_events.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+class AllEventsSchema(Schema, ABC):
+    def get_data_source(self) -> RelationalSource:
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return get_dataset(EVENTS) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: Not sure I follow this:
this is supposed to be the abstract schema for the virtual table sentry will see. Shouldn't it have all columns coming from both transactions and events together (with the nullability tweaked accordingly) ?

File: snuba/datasets/all_events.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    if [col for col in TRANSACTIONS_ONLY_COLUMNS if col in get_condition_columns(query)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class AllEventsProcessor(QueryProcessor):
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: This processor will become big. I think you should:
1) move it to its own file
2) Give it a more specific name. We are likely going to  have multiple processors per dataset. So they should provide some context about what they do. This specifically switches table (people would expect it changes the FROM clause if we name it appropriately).

File: snuba/datasets/all_events.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            .get_read_schema() \
+            .get_data_source()
+        query.set_data_source(source)
+
+        if detected_dataset == TRANSACTIONS:
+            with_columns = [
+                ("'transaction'", 'type'),
+                ('finish_ts', 'timestamp'),
+                ('user_name', 'username'),
+                ('user_email', 'email'),
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: what's the advantage of using a with instead of just replacing the columns in the query object itself (besides the fact that I did not yet provide a conveniente `replace_column_in_query` method to use and it has to be done manually)? 
Going on, replacing the column names is going to be a more reliable solution since it leaves a query that is easier to process for the next processor. Using the WITH clause will require the next processor to understand the WITH and figure out what changes to do.

File: snuba/datasets/all_events.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    Experimental dataset for Discover
+    that coerces the columns of Events and Transactions into the same format
+    and sends a query to either one.
+
+    Currently does this by switching between events and transactions tables
+    depending on the conditions in the provided query.
+    """
+    EVENTS_ALIAS = 'events'
+    TRANSACTIONS_ALIAS = 'transactions'
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: Never used ?

File: snuba/datasets/all_events.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+def get_condition_columns(query: Query) -> Sequence[Any]:
+    col_exprs: MutableSequence[Any] = []
+    # Conditions need flattening as they can be nested as AND/OR
+    if query.get_conditions():
+        flat_conditions = list(
+            chain(*[[c] if is_condition(c) else c for c in query.get_conditions()]))
+        col_exprs.extend([c[0] for c in flat_conditions])
+
+    # Return the set of all columns referenced in any expression
+    return set(chain(*[columns_in_expr(ex) for ex in col_exprs]))
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: Feel free to make changes to the Query api to make get_all_referenced_columns more flexible. Something like this will be useful to others as well. (This does not mean "just copy this method back into Query". Separating the logic that iterates over columns and returns them in its own method and adding public methods that iterate over specific parts of the query (columns, conditions, etc.) seems a more flexible solution with little code duplication.

File: snuba/datasets/all_events.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        query.set_with(with_columns)
+
+
+class AllEventsSchema(Schema, ABC):
+    def get_data_source(self) -> RelationalSource:
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return get_dataset(EVENTS) \
```
Comment (fpacifici) [inline snuba/datasets/all_events.py]: This is a standard table relational source, which means bootstrap will build the table in clickhouse as soon as you try to use it. Now it probably did not happen since the schema you return is the Events one.
I think the issue tight now is that there are only two categories of RelationalSource: Tables (whith DDL instructions) and JoinSources (which join multiple tables). You do not want any of them what you are looking into is a new subclass of RelationalSchema, that resembles table but that only provides a read schema that does not change and does not get built by bootstrap + a way to swap the real table used to generate the FROM clause.
I suspect this should stay in its own PR since while not complex conceptually, it is going to be verbose and it is going to touch a lot of different places.

File: snuba/datasets/all_events.py (hunk 1 above)
Comment (lynnagara) [reply snuba/datasets/all_events.py]: Alternate names I considered:
- DiscoverDataset
- (Merged)EventsAndTransactionsDataset

Comment (fpacifici) [reply snuba/datasets/all_events.py]: Cannot think about a good one but I think DiscoverDataset is more representative of what this dataset does than AllEvents.

File: snuba/datasets/all_events.py (hunk 2 above)
Comment (lynnagara) [reply snuba/datasets/all_events.py]: I changed this to just return None. In the current implementation I wanted to avoid introducing a new data source since this dataset only switches between the existing events and transactions sources right now. These lines here are just placeholder, since the dataset actually gets picked in the processor depending on what dataset is detected.

Are you proposing doing the dataset detection inside DataSource instead of the query processor and then making DataSource pick what table gets selected?

Comment (fpacifici) [reply snuba/datasets/all_events.py]: We still need an abstract schema (see the get_columns method).
While it is true that today we would not be using it, soon we will use it to parse the column queries and to validate the query itself.
Also it would document how this dataset works and what it returns, otherwise people will have to guess by reading the processors.

>Are you proposing doing the dataset detection inside DataSource instead of the query processor and then making DataSource pick what table gets selected?
No, please let's keep this in the processor. Or at least, if you want to put the decision logic in the DataSource I am fine as long as it is triggered by a query processor so we respect the rule based abstraction. But in that case such change would not be very useful.

File: snuba/datasets/all_events.py (hunk 4 above)
Comment (lynnagara) [reply snuba/datasets/all_events.py]: We can use `column_expr` to achieve this today even before we get the new method. However there are some edge cases that WITH solves that I need to verify for the column replacement solution. OTOH these will be things like-
- columns like `uniq(colname)`
- column not referenced anywhere except in orderby
- having clause

File: snuba/datasets/all_events.py (hunk 2 above)
Comment (lynnagara) [reply snuba/datasets/all_events.py]: No I definitely would prefer to keep it in the processor. I was just wondering why you were proposing to provide a fully detailed DataSource that is not used. If it's just for documentation now that's fine.

File: snuba/datasets/discover.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    'start_ts',
+    'start_ms',
+    'finish_ts',
+    'finish_ms',
+    'duration',
+]
+
+EVENTS_ONLY_COLUMNS = [
+    'group_id',
+    'transaction',
```
Comment (markstory) [inline snuba/datasets/discover.py]: What about the various promoted tag columns?

File: snuba/datasets/all_events.py (hunk 4 above)
Comment (fpacifici) [reply snuba/datasets/all_events.py]: I'd advise against using column_expr is not strictly necessary since it works at a different point in the query processing (during formatting basically), so it will be harder to move that back.
About the column above #555 should solve them. If you need a stop gap solution, ok, let's add them to column_expr until the PR is committed, but let's ensure 555 works for you so we can move them back easily.

About the HAVING clause. I think we can support it in #555. My theory is that it was just forgotten from all_referenced_columns. It is a quick fix

'column not referenced anywhere except in orderby' what's the problem here?

`uniq(colname)` I think you will still see such a column in all_referenced_column and can replace it in #555. Do you have a specific example where this does not work ?

Comment (lynnagara) [reply snuba/datasets/all_events.py]: Sorry, it was just a note to self of some cases I wanted to remember to test against the column_expr implementation. Seems to all be working fine.

File: snuba/datasets/discover.py (hunk 8 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Thanks, they should be added now

File: snuba/datasets/discover.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            ('ip_address', Nullable(String())),
+        ])
+
+        return common + self.get_events_only_columns() + self.get_transactions_only_columns()
+
+    def get_events_only_columns(self) -> ColumnSet:
+        return ColumnSet([
+            ('group_id', Nullable(UInt(64))),
+            ('primary_hash', Nullable(FixedString(32))),
+            ('type', Nullable(String())),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Should we consider this common and auto-fill it with "transaction" for transactions ? Which also makes it non nullable.

File: snuba/datasets/discover.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            # Promoted tags
+            ('level', Nullable(String())),
+            ('logger', Nullable(String())),
+            ('server_name', Nullable(String())),
+            ('transaction', Nullable(String())),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Is this related with transaction_name ?

File: snuba/datasets/discover.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+            ('exception_stacks.mechanism_handled', Nullable(UInt(8))),
+            ('exception_frames.abs_path', Nullable(String())),
+            ('exception_frames.filename', Nullable(String())),
+            ('exception_frames.package', Nullable(String())),
+            ('exception_frames.module', Nullable(String())),
+            ('exception_frames.function', Nullable(String())),
+            ('exception_frames.in_app', Nullable(UInt(8))),
+            ('exception_frames.colno', Nullable(UInt(32))),
+            ('exception_frames.lineno', Nullable(UInt(32))),
+            ('exception_frames.stack_level', Nullable(UInt(16))),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: This is a nested column on events. Is this intentional ?

File: snuba/datasets/discover.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        return ColumnSet([
+            ('trace_id', Nullable(UUID())),
+            ('span_id', Nullable(UInt(64))),
+            ('transaction_name', String()),
+            ('transaction_hash', Nullable(UInt(64))),
+            ('transaction_op', Nullable(String())),
+            ('start_ts', Nullable(DateTime())),
+            ('start_ms', Nullable(UInt(16))),
+            ('finish_ts', Nullable(DateTime())),
+            ('finish_ms', Nullable(UInt(16))),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: This is the same as timestamp on the common fields. It is up to us to decide whether to format that as datetime or as two fields

File: snuba/datasets/discover.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+class DiscoverSchema(Schema, ABC):
+    def get_data_source(self):
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return None
+
+    def get_columns(self) -> ColumnSet:
+        common = ColumnSet([
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: I think we are missing tags and contexts

File: snuba/datasets/discover.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+class DiscoverSchema(Schema, ABC):
+    def get_data_source(self):
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return None
+
+    def get_columns(self) -> ColumnSet:
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: @markstory  your input will be useful about the data model below. A lot of this can be adjusted to make it simpler to deal with the dataset on Sentry.

File: snuba/datasets/discover.py [hunk 15]
```diff
@@ -85,22 +86,54 @@ def get_data_source(self):
...
+        common = ColumnSet([
+            ('event_id', FixedString(32)),
+            ('project_id', UInt(64)),
+            ('timestamp', DateTime()),
+            ('deleted', UInt(8)),
+            ('retention_days', UInt(16)),
+            ('platform', Nullable(String())),
+            ('sentry:release', Nullable(String())),
+            ('sentry:dist', Nullable(String())),
+            ('sentry:user', Nullable(String())),
```
Comment (markstory) [inline snuba/datasets/discover.py]: Why these names over `release`, `dist` and `user`? Should `environment` be here too?

File: snuba/datasets/discover.py [hunk 16]
```diff
@@ -85,22 +86,54 @@ def get_data_source(self):
         return None
 
     def get_columns(self) -> ColumnSet:
-        common = ColumnSet([])  # TODO: Fill out
+        common = ColumnSet([
+            ('event_id', FixedString(32)),
+            ('project_id', UInt(64)),
+            ('timestamp', DateTime()),
```
Comment (markstory) [inline snuba/datasets/discover.py]: Will this map to `finish_ts` for transaction events?

File: snuba/datasets/discover.py [hunk 17]
```diff
@@ -85,22 +86,54 @@ def get_data_source(self):
...
+            ('retention_days', UInt(16)),
+            ('platform', Nullable(String())),
+            ('sentry:release', Nullable(String())),
+            ('sentry:dist', Nullable(String())),
+            ('sentry:user', Nullable(String())),
+            # User
+            ('user_id', Nullable(String())),
+            ('username', Nullable(String())),
+            ('email', Nullable(String())),
+            ('ip_address', Nullable(String())),
```
Comment (markstory) [inline snuba/datasets/discover.py]: Does this coalesce the v4 and v6 columns for transactions?

File: snuba/datasets/discover.py (hunk 9 above)
Comment (markstory) [reply snuba/datasets/discover.py]: Having it synthesized for transactions would be great. Makes the result sets appear more consistent to the end user.

File: snuba/datasets/discover.py (hunk 10 above)
Comment (markstory) [reply snuba/datasets/discover.py]: It should be. In the current external schema they are the same thing.

File: snuba/datasets/discover.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+        return ColumnSet([
+            ('group_id', Nullable(UInt(64))),
+            ('primary_hash', Nullable(FixedString(32))),
+            ('type', Nullable(String())),
+            # Promoted tags
+            ('level', Nullable(String())),
+            ('logger', Nullable(String())),
+            ('server_name', Nullable(String())),
+            ('transaction', Nullable(String())),
+            ('site', Nullable(String())),
```
Comment (markstory) [inline snuba/datasets/discover.py]: I don't see this in the current events schema.

File: snuba/datasets/discover.py (hunk 12 above)
Comment (markstory) [reply snuba/datasets/discover.py]: Getting a single datetime value is more useful for sentry.

File: snuba/datasets/discover.py [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+            ),
+        }
+
+    def column_expr(self, column_name, query: Query, parsing_context: ParsingContext):
+        detected_dataset = detect_dataset(query)
+
+        if detected_dataset == TRANSACTIONS:
+            if column_name == 'type':
+                return "'transaction'"
+            if column_name == 'timestamp':
```
Comment (markstory) [inline snuba/datasets/discover.py]: There is also the time group alias that is used in timeseries results. It currently maps to the `bucketed_end` alias or `time` for events.

File: snuba/datasets/discover.py (hunk 9 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Thanks made it not nullable. It's already being auto filled

File: snuba/datasets/discover.py (hunk 12 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Timestamp is being filled with the `finish_ts` value. I left `finish_ts` in as well though for transactions.

File: snuba/datasets/discover.py (hunk 10 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Moved to common columns and filled from transaction_name for the transactions dataset

File: snuba/datasets/discover.py (hunk 11 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Yeah, i think it's easier if the Discover dataset doesn't need to know about columns that might be nested in one of the datasets. It's going to be a lot easier to map and fill in values and types if columns are not nested. More immediately `ColumnSet.get()` won't work when we pass the expanded column name on nested types.

File: snuba/datasets/discover.py (hunk 15 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Yes, added `environment`. I picked these names since they are the ones in the Events dataset and those are the names Sentry is always passing. Introducing completely new names while we do the migration is going to be complicated, and Sentry maps them to other public names anyway.

File: snuba/datasets/discover.py (hunk 12 above)
Comment (fpacifici) [reply snuba/datasets/discover.py]: What's the reason for keeping finish_ts ?

To give some context:
- the precision of clickhouse datetime columns is "seconds"
- that is why we added separate columns finish_ts/finish_ms 
- datetime in python do not have this constraint, we should not need two separate representations.

File: snuba/datasets/discover.py (hunk 16 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Yes, this is done in `column_expr`

File: snuba/datasets/discover.py (hunk 17 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Good call, added this for both the transactions and Discover datasets

File: snuba/datasets/discover.py (hunk 18 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: It's in the promoted tags list

File: snuba/datasets/discover.py (hunk 13 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Added

File: snuba/datasets/discover.py (hunk 12 above)
Comment (fpacifici) [reply snuba/datasets/discover.py]: @markstory Thinking over this, actually having a single datetime is not something we can do in the dataset definition, we need to work on the snuba reader for that. /snuba/reader.py. We should not try to achieve that in this PR, it is going to be complex.  This is because we do not have a dataset specific resultProcessor.
So I would keep timestamp + milliseconds as an int.

@lynnagara, I don't think we should leave the finish_ts field. (or maybe keep only the finish_ts). Duplicated columns are going to be confusing.

File: snuba/datasets/discover.py (hunk 19 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: I added `time` and `bucketed_end` to the time columns. It would be nice if we could just use `time` for both, but this might require some larger changes.

File: snuba/datasets/discover.py (hunk 9 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Actually, now I remember why it's nullable. It's because it's defined as nullable in Events. Changing it back.

File: snuba/datasets/discover.py (hunk 12 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Because of the way that time groupings work both fields are still referenced there. We have `time` and `bucketed_end` right now. It would be better to just have `time` but I don't have an easy fix for this.

File: snuba/datasets/discover.py (hunk 9 above)
Comment (fpacifici) [reply snuba/datasets/discover.py]: meh

File: snuba/datasets/discover.py (hunk 12 above)
Comment (fpacifici) [reply snuba/datasets/discover.py]: so the issue is that time group columns are managed in column_expr where we do not know much about dataset (technically we have access to the Query there, so we could do it properly). Anyway, why don't we define two grouping, one for finish_ts and one for timestamp, run a query processor first that would turn a grouped timestamp column into one of the two which will then be resovled by column_Expr ?

File: snuba/datasets/discover.py [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: I think the "having" clause is relevant as well.

File: snuba/datasets/discover.py [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: conditions can be nested:
```
[
    [
        ['w', '=', '1'], ['x', '=', '2']
    ],
    [
        ['y', '=', '3'], ['z', '=', '4']
    ],
]

(w = '1' OR x = '2') AND (y = '3' OR z = '4')
```

File: snuba/datasets/discover.py [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    transaction_column_set = get_dataset('discover') \
+        .get_dataset_schemas() \
+        .get_read_schema() \
+        .get_transactions_only_columns()
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Please pass the column sets this needs in the constructor. While right now it is possible to load a dataset from here, this is a coincidence of the fact that get_dataset loads dependencies inline, but that may not last. Depending on a dataset from here is a circular dependency even if python does notice it.

File: snuba/datasets/discover.py [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+        query.set_data_source(source)
+
+
+class DiscoverSchema(Schema, ABC):
+    def get_data_source(self):
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return None
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: This won't type check. If you need an empty RelationalSource that needs additional input to switch from a table to another you will need to create a class for that, that extends RelationalSource

File: snuba/datasets/discover.py [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+
+    def get_columns(self) -> ColumnSet:
+        common = ColumnSet([
+            ('event_id', FixedString(32)),
+            ('project_id', UInt(64)),
+            ('timestamp', DateTime()),
+            ('deleted', UInt(8)),
+            ('retention_days', UInt(16)),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: You do not need these columns. This is for clickhouse and we do not need to expose it in the virtual table.

File: snuba/datasets/transactions.py [hunk 25]
```diff
@@ -182,6 +182,8 @@ def column_expr(self, column_name, query: Query, parsing_context: ParsingContext
             return 'IPv4NumToString(ip_address_v4)'
         if column_name == 'ip_address_v6':
             return 'IPv6NumToString(ip_address_v6)'
+        if column_name == 'ip_address':
+            return f"coalesce(IPv4NumToString(ip_address_v4), IPv6NumToString(ip_address_v6))"
```
Comment (fpacifici) [inline snuba/datasets/transactions.py]: you do not need IPv6NumToString and IPv4NumToString. We upgraded the driver to access clickhouse.

File: snuba/datasets/discover.py [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+            ),
+            'timeseries': TimeSeriesExtension(
+                default_granularity=3600,
+                default_window=timedelta(days=5),
+                timestamp_column='timestamp',
+            ),
+        }
+
+    def column_expr(self, column_name, query: Query, parsing_context: ParsingContext):
+        detected_dataset = detect_dataset(query)
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: you need to delegate the tags processing to TagColumnProcessor otherwise they will not be resolved.

Comment (fpacifici) [reply snuba/datasets/discover.py]: Also could you please depend on the query object to understand which table we are querying `get_data_source`.
The content of the query may have been mutated by the time you get here. The result of detect dataset may not be valid anymore.

File: snuba/datasets/discover.py [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+                return 'user_name'
+            if column_name == 'email':
+                return 'user_email'
+            if column_name == 'transaction':
+                return 'transaction_name'
+            if self.get_dataset_schemas() \
+                    .get_read_schema() \
+                    .get_events_only_columns() \
+                    .get(column_name):
+                return 'NULL'
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Does this query works if I run something like 
select duration, count(*) FROM sentry where type != transaction group by duration

File: snuba/query/query.py [hunk 28]
```diff
@@ -217,6 +217,7 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
             col_exprs.extend(to_list(self.get_selected_columns()))
 
         # Conditions need flattening as they can be nested as AND/OR
+        self.__add_flat_conditions(col_exprs)
         if self.get_conditions():
             flat_conditions = list(chain(*[[c] if is_condition(c) else c for c in self.get_conditions()]))
             col_exprs.extend([c[0] for c in flat_conditions])
```
Comment (fpacifici) [inline snuba/query/query.py]: This is now duplicated

File: snuba/datasets/discover.py (hunk 20 above)
Comment (markstory) [reply snuba/datasets/discover.py]: Yes, we'll be using `having` soon enough in discover2.

File: snuba/datasets/discover.py (hunk 19 above)
Comment (markstory) [reply snuba/datasets/discover.py]: I'm :+1: in using `time` as the standard alias across both datasets.

File: snuba/datasets/discover.py (hunk 12 above)
Comment (markstory) [reply snuba/datasets/discover.py]: The bucketed `time` grouping column doesn't need ms precision as we're always grouping on wider ranges.

We currently don't read the `finish_ms` directly out of snuba. The only time we use the millisecond precision is when we're rendering the span graph but all of that data is already coming out of nodestore. That is a roundabout way of saying I rescind my earlier comment and am ok with separate timestamp and ms columns.

File: snuba/datasets/discover.py (hunk 20 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Tackled in https://github.com/getsentry/snuba/pull/561

File: snuba/datasets/discover.py (hunk 21 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: This is a very limited dataset detection function, which matches the logic that is in Sentry right now. If there is a nested OR condition on type, we still cannot be sure of the dataset so we do not handle that case in this way.

Comment (fpacifici) [reply snuba/datasets/discover.py]: fair point.

File: snuba/datasets/discover.py (hunk 27 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Yes it yields a result that looks like `[{"duration": None, "count": 999}]`

File: snuba/datasets/discover.py (hunk 22 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Hmm, yeah. Fixed this up.

File: snuba/datasets/discover.py (hunk 26 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Currently TagColumnProcessor gets called via `EventsDataset.column_expr` or `TransactionsDataset.column_expr`. Is there a reason we'd rather have this logic in the Discover dataset?

Comment (fpacifici) [reply snuba/datasets/discover.py]: I did not notice you were delegating the call to the events and transactions dataset `column_expr` method, which is something we will have eventually to remove. 
Ok, ignore the comment about tags. Please do not depend on detect_dataset here though. This call happens after all query processors have already run, so the query itself may be different.

File: snuba/datasets/discover.py [hunk 29]
```diff
@@ -0,0 +1,278 @@
...
+class DiscoverSource(RelationalSource):
+    def __init__(self, columns: ColumnSet, discover_table_source: DiscoverTableSource) -> None:
+        self.__columns = columns
+        self.__discover_table_source = discover_table_source
+
+    def format_from(self) -> str:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.format_from()
+        return ""
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: If this is not a valid case let's raise an error

File: snuba/datasets/discover.py [hunk 30]
```diff
@@ -0,0 +1,278 @@
...
+
+    def get_columns(self) -> ColumnSet:
+        return self.__columns
+
+    def get_mandatory_conditions(self) -> Sequence[Condition]:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.get_mandatory_conditions()
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: :+1:

File: snuba/datasets/discover.py [hunk 31]
```diff
@@ -0,0 +1,278 @@
...
+        return []
+
+
+class DiscoverProcessor(QueryProcessor):
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Processors should be relatively simple rules applied in sequence to a query. Could you please give a more specific name to this one, since likely there won't be a single big processor for Discover.

File: snuba/datasets/discover.py [hunk 32]
```diff
@@ -0,0 +1,278 @@
...
+    the detected dataset.
+    """
+
+    def __init__(self, table_source: DiscoverTableSource, transactions_columns: ColumnSet) -> None:
+        self.__table_source = table_source
+        self.__transactions_columns = transactions_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        detected_dataset = detect_dataset(query, self.__transactions_columns)
+        self.__table_source.set_table_source(detected_dataset)
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Please di not store mutable state in the processor. There is no guarantee instances will not be reused:
https://github.com/getsentry/snuba/blob/master/snuba/query/query_processor.py#L20-L21

__transactions_columns is fine because you are not mutating that state

Why aren't you just updating the Query ?

Comment (lynnagara) [reply snuba/datasets/discover.py]: So I wanted this to be something like `query.get_data_source().set_table_source()` but I haven't come up with a way to do this that satisfies the type checker

File: snuba/datasets/discover.py [hunk 33]
```diff
@@ -0,0 +1,301 @@
...
+        if transactions_columns.get(col)
+    ]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class DiscoverTableSource:
+    def __init__(self):
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Please add the missing return type hint
> def __init__(self) -> None:

File: snuba/datasets/discover.py [hunk 34]
```diff
@@ -0,0 +1,301 @@
...
+    ) -> None:
+        self.__discover_source = discover_source
+        self.__transactions_columns = transactions_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        detected_dataset = detect_dataset(query, self.__transactions_columns)
+        table_source = DiscoverTableSource()
+        table_source.set_table_source(detected_dataset)
+        data_source = DiscoverSource(self.__discover_source.get_columns(), table_source)
+        query.set_data_source(data_source)
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Any reason for not just making DiscoverSource mutable itself?

```
source = query.get_data_source()
assert isinstance(source, DiscoverSource) # this is a filter for discover it makes sense it does not work for other sources
source.set_table(table_source)
```

File: snuba/datasets/discover.py [hunk 35]
```diff
@@ -0,0 +1,301 @@
...
+        self.__transactions_columns = ColumnSet(
+            [
+                ("trace_id", Nullable(UUID())),
+                ("span_id", Nullable(UInt(64))),
+                ("transaction_hash", Nullable(UInt(64))),
+                ("transaction_op", Nullable(String())),
+                ("start_ts", Nullable(DateTime())),
+                ("start_ms", Nullable(UInt(16))),
+                ("finish_ts", Nullable(DateTime())),
+                ("finish_ms", Nullable(UInt(16))),
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: please let's have a TODO about cleaning this up once we have fixed how the timeseries works.

File: snuba/datasets/discover.py (hunk 34 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Oh interesting. I didn't realise that adding the assertion would satisfy the typechecker. Gonna update to this then.

File: snuba/datasets/discover.py (hunk 33 above)
Comment (lynnagara) [reply snuba/datasets/discover.py]: Removed this class completely
